src/main/perf/LineFileDocs.java

```diff
@@ -745,28 +751,71 @@ public Document nextDoc(DocState doc) throws IOException {
     }
   }
 
-  private static class LineFileDoc {
-    final FloatBuffer vector;
-    final String stringText;
-    final ByteBuffer byteText;
+  private static abstract class LineFileDoc {
+    private final FloatBuffer vector;
 
-    LineFileDoc(String text, float[] vector) {
-      stringText = text;
-      byteText = null;
+    LineFileDoc(float[] vector) {
       if (vector == null) {
         this.vector = null;
       } else {
         this.vector = FloatBuffer.wrap(vector);
       }
     }
 
-    LineFileDoc(ByteBuffer bytes, float[] vector) {
-      stringText = null;
-      byteText = bytes;
-      if (vector == null) {
-        this.vector = null;
-      } else {
-        this.vector = FloatBuffer.wrap(vector);
+    abstract int getId();
+
+    String getStringText() {
+      throw new UnsupportedOperationException();
+    }
+
+    ByteBuffer getByteText() {
+      throw new UnsupportedOperationException();
+    }
+
+    private static final class TextBased extends LineFileDoc {
+      final String stringText;
+
+      final int id; // will be the exact id if this is txt based LFD
+      TextBased(String text, float[] vector, int id) {
+        super(vector);
+        stringText = text;
+        this.id = id;
+      }
+
+      @Override
+      int getId() {
+        return id;
+      }
+
+      @Override
+      String getStringText() {
+        return stringText;
+      }
+    }
+
+    private static final class BinaryBased extends LineFileDoc {
+
+      final ByteBuffer byteText;
+      private int nextId; // will have multiple doc in a same LFD so we'll determine id using a base
+      private int docCount; // and a count
+      BinaryBased(ByteBuffer bytes, float[] vector, int idBase, int docCount) {
+        super(vector);
+        byteText = bytes;
+        this.nextId = idBase;
+        this.docCount = docCount;
+      }
+
+      @Override
+      int getId() {
+        if (docCount-- == 0) {
```

Comment:
Hmm this is a little too sneaky -- it looks like a getter but it has this subtle side-effect.

Reply:
I'll rename it to `getNextId`? Then it won't be that weird if there's some exception throwing out?

Reply:
Thanks, this is better.

But: I am still confused.  Won't this mean we increment over N docs (ids) in one block, but index exactly the same text and vector each time?

Reply:
Oh!  I see now!  The text and vector are concatenated values for all the docs in this block!

OK, phew.

Can you rename `byteText` to something like `blockByteText`?   I guess for the vector we can leave as is since it's stored in base class.  But can you add a comment to both explaining that these are appended values for all docs in the block?

Reply:
Sure I added some comments to help understand the code :)